### PR TITLE
improves sigverify discard_excess_packets performance

### PR DESCRIPTION
**Pull request by behzadnouri**
#### Problem
As shown by the added benchmark, current code does worse if there is a
spam address plus a lot of unique addresses.

on current master:
```
test bench_packet_discard_many_senders  ... bench:   1,997,960 ns/iter (+/- 103,715)
test bench_packet_discard_mixed_senders ... bench:  14,256,116 ns/iter (+/- 534,865)
test bench_packet_discard_single_sender ... bench:   1,306,809 ns/iter (+/- 61,992)
```

with this commit:
```
test bench_packet_discard_many_senders  ... bench:   1,644,025 ns/iter (+/- 83,715)
test bench_packet_discard_mixed_senders ... bench:   1,089,789 ns/iter (+/- 86,324)
test bench_packet_discard_single_sender ... bench:     955,234 ns/iter (+/- 55,953)
```
#### Summary of Changes
* No change to the order of packets or which packets are discarded.
* Drop hash-map entries if there are no more packets from respective address.
* Allocate `max_packets` with fewer iterations.

### CODE REVIEW

**Review by sakridge — APPROVED** [review]
lgtm

**Comment by buffalu** [comment]
fuck yeah

File: core/src/sigverify_stage.rs [hunk 1]
```diff
@@ -161,34 +161,27 @@ impl SigVerifyStage {
...
+            .rev()
+            .flat_map(|batch| batch.packets.iter_mut().rev())
+            .map(|packet| (packet.meta.addr, packet))
+            .into_group_map();
+        // Allocate max_packets evenly across addresses.
+        while max_packets > 0 && !addrs.is_empty() {
+            let num_addrs = addrs.len();
+            addrs.retain(|_, packets| {
+                let cap = (max_packets + num_addrs - 1) / num_addrs;
+                max_packets -= packets.len().min(cap);
```
Comment (jstarry) [inline core/src/sigverify_stage.rs:177]: Couldn't this underflow?

Comment (behzadnouri) [reply core/src/sigverify_stage.rs:177]: I don't think so. Do you have a particular example?

**Pull request merged**